File: src/main/java/com/HOIIVUtils/clauzewitz/script/BooleanPDX.java
package com.HOIIVUtils.clauzewitz.script;

import com.HOIIVUtils.clauzewitz.BoolType;
import com.HOIIVUtils.clausewitz_parser.Node;
import com.HOIIVUtils.clausewitz_parser.NodeValue;

import java.util.List;

public class BooleanPDX extends AbstractPDX<Boolean> {
    private final BoolType boolType;
    /**
     * this is separate to obj because
     * if we want to clear the value of obj, we would still return this value by default.
     */
    private final boolean defaultValue;

    public BooleanPDX(String pdxIdentifiers, boolean defaultValue, BoolType boolType) {
        super(pdxIdentifiers);
        this.boolType = boolType;
        this.defaultValue = defaultValue;
    }

    public BooleanPDX(List<String> pdxIdentifiers, boolean defaultValue, BoolType boolType) {
        super(pdxIdentifiers);
        this.boolType = boolType;
        this.defaultValue = defaultValue;
    }

    @Override
    public void set(Node expression) throws UnexpectedIdentifierException, NodeValueTypeException {
        usingIdentifier(expression);
        NodeValue value = expression.value();

        if (value.valueObject() instanceof String str) {
            obj = value.bool(boolType);
        } else {
            throw new NodeValueTypeException(expression, "String parsable as Bool matching enum + " + boolType.toString());
        }
    }

    @Override
    public boolean objEquals(PDXScript<?> other) {
        if (other instanceof BooleanPDX pdx) {
            return obj.equals(pdx.get());
        }
        return false;
    }
}

File: src/main/scala/com/hoi4utils/script/BooleanPDX.scala
package com.hoi4utils.script

import com.hoi4utils.BoolType
import com.hoi4utils.exceptions.{NodeValueTypeException, UnexpectedIdentifierException}
import com.hoi4utils.parser.Node

class BooleanPDX(pdxIdentifiers: List[String], final private var defaultValue: Boolean, final private var boolType: BoolType) extends AbstractPDX[Boolean](pdxIdentifiers) {

  def this(pdxIdentifier: String, defaultValue: Boolean, boolType: BoolType) = {
    this(List(pdxIdentifier), defaultValue, boolType)
  }

  @throws[UnexpectedIdentifierException]
  @throws[NodeValueTypeException]
  override def set(expression: Node): Unit =
    usingIdentifier(expression)
    this.node = Some(expression)
    expression.$ match
      case _: Boolean => // No action needed, value is already Boolean
      case _: String =>
        val v = this.node.get.$.asInstanceOf[String]
        if boolType.matches(v) then set(boolType.parse(v))
        else throw NodeValueTypeException(expression, s"String parsable as Bool matching enum + ${boolType.toString}", s"${expression.$}")
      case _ =>
        throw NodeValueTypeException(expression, "Boolean or String", s"${expression.$}")
  
  override def set(value: Boolean): Boolean = {
    if (this.node.nonEmpty)
      this.node.foreach(_.setValue(value))
    else
      this.node = Some(Node(value))
    value
  }

  override def value: Option[Boolean] = {
    val v = super.value
    v.orNull match {
      case b: Boolean => Some(b)
      case null => None
    }
  }

  def $ : Boolean = {
    value.getOrElse(defaultValue)
  }

  def objEquals(other: PDXScript[?]): Boolean = {
    other match {
      case b: BooleanPDX => node.equals(b.node)
      case _ => false
    }
  }

  def invert: Boolean = {
    setNode(!this.$)
    this.$
  }

  override def toString: String = this.value.map(_.toString).getOrElse("")
}
